File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool sameTree(TreeNode* a, TreeNode* b){
        if(!a && !b){
            return true;
        }else if(!a){
            return false;
        }else if(!b){
            return false;
        }
        if(a->val == b->val){
            return sameTree(a->left, b->left ) && sameTree(a->right, b->right);
        }
        return false;
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!root && !subRoot){
            return true;
        }else if(!root){
            return false;
        }else if(!subRoot){
            return false;
        }
        if(sameTree(root, subRoot)){
            return true;
        }else{
            return isSubtree(root->left , subRoot) ||  isSubtree(root->right , subRoot);
        }
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = [root.val]
        # DFS + Inorder
        def dfs(root):
            if not root:
                return 0

            leftmax = dfs(root.left)
            rightmax = dfs(root.right)

            leftmax = max(leftmax, 0)
            rightmax = max(rightmax, 0)

            if leftmax + rightmax + root.val > res[0]:
                res[0] = leftmax + rightmax + root.val
            # need to return the part of larger branch 
            return max(leftmax, rightmax) + root.val
            
        dfs(root)
        return res[0]